feat: remove server peer

RangerMauve opened the pull request:
Closes #999

Some specific stuff I'd like feedback on:
1. Right now this only allows removing while we're replicating. Should I initialize a replication stream if one doesn't exist instead of erroring?
2. Is removing the server by the base URL the easiest for the front end? Should we use the deviceId or some other identifier instead?
3. Are there any events we should be adding to let the front end know the server is removed, or is the role update for the server peer enough?
4. Any other tests I should include here?

achou11 commented:
Haven't looked at the PR just yet but answering where I can...

---

> Should I initialize a replication stream if one doesn't exist instead of erroring?

I think so? From what I remember, we have 2 kinds of sync: "initial" and "data" sync. In theory, each one could be enabled/disabled at will, but in practice I think we always have the "initial" sync enabled and allow the "data" sync to be toggled (see `SyncApi.start()` and `SyncApi.stop()`). Removal of a remote archive from a project affects the `auth` namespace, which falls under the "initial" sync. If we don't have anything exposed to the client in terms of starting/stopping the "initial" sync, not sure what actions the client could take if you throw an error when attempting to remove a server.

Apologies if this isn't relevant to your question! "replicating" feels like it could mean a bunch of different things and this was the thing that first came to mind.

--- 

> Right now this only allows removing while we're replicating.

Does this detail from the original issue conflict with this implementation detail? 

> should handle the edge-case of the coordinator writing the new "blocked" role for the server, but not having a connection to the server at that time, or loosing the connection before syncing. The device that removes the server should keep trying to connect until it confirms that the blocked role record has synced to the server. This retry should persist across app restarts.

Could be tangential but just wanted to check.

---

> Is removing the server by the base URL the easiest for the front end? Should we use the deviceId or some other identifier instead?

I _think_ `deviceId` would be slightly easier for the frontend since remote archives are project members. The general sequence from the frontend would be:

  1. Get all members of the project
  2. Filter out the non-archive members
  3. Select one of the remaining members and call `project.$member.removeServerPeer(member.deviceId)`
  
To me, it'd be nice if we could get closer to something like `MemberApi.remove(deviceId)`, where it only actually does something for remote archives for now, but could eventually be used for non-archive members too (if that's a future need we have).

---

> Are there any events we should be adding to let the front end know the server is removed, or is the role update for the server peer enough?

I think the role update for the server peer is sufficient for our needs. After removing the server, we'd do a refetch of the project members and see that the role has been updated that way.

RangerMauve commented:
Very useful feedback, ty. Going to do some refactoring based on it. I agree `.remove()` could be handy down the line.

Regarding the edge case, this will happen in a future PR. Going to handle the online case first in order to unblock folks, then add resilience to remove while offline later

RangerMauve commented:
Simplified the code a bunch.

- fixed `MembersAPI.getById` not supporting the self hosted server device info
- We always establish a connection on remove and wait for initial sync, bypassing the rest of the sync API.
- Added a test to make sure the server knows that it is blocked after we sync

achou11 commented on src/member-api.js:
```diff
@@ -312,6 +317,46 @@ export class MemberApi extends TypedEmitter {
+    { dangerouslyAllowInsecureConnections = false } = {}
+  ) {
+    // Get device ID for URL
+    // Prase through URL to ensure end pathname if missing
```
parse?
```suggestion
    // Parse through URL to ensure end pathname if missing
```
